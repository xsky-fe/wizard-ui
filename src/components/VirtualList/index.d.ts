import React from 'react';
import { VirtualListProps, VirtualListState } from '../../interface';

declare class VirtualList<T> extends React.Component<VirtualListProps<T>, VirtualListState>{ }

export default VirtualList;