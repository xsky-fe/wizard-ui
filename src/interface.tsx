import * as React from 'react';
import {
  SelectCallback,
  Sizes,
} from 'react-bootstrap';
import { Moment } from 'moment';
import CSS from 'csstype';

export interface Map<K, V> {
  clear(): void;
  delete(key: K): boolean;
  entries(): IterableIterator<[K, V]>;
  forEach(callbackfn: (value: V, index: K, map: Map<K, V>) => void, thisArg?: any): void;
  get(key: K): V;
  has(key: K): boolean;
  keys(): IterableIterator<K>;
  set(key: K, value?: V): Map<K, V>;
  size: number;
  values(): IterableIterator<V>;
  [Symbol.iterator](): IterableIterator<[K, V]>;
  [Symbol.toStringTag]: string;
}

export interface Query {
  offset?: number;
  limit?: number;
  q?: string;
}

export interface FetchResponse<T = any> {
  response?: {
    [res: string]: T | T[];
  };
  error?: string;
}

export interface BadgeProps {
  count?: number | string;
  showZero?: boolean;
  dot?: boolean;
  size?: string;
  status?: string;
  overflowCount?: number;
  text?: string;
  title?: string;
}

export interface IconProps {
  type: string;
  color?: string;
  className?: string;
  onClick?: React.MouseEventHandler;
  style?: object;
}

export interface TooltipProps {
  iconClass?: string;
  icon?: string;
  iconAlign?: string;
  onClick?: any;
  label?: React.ReactNode;
  contrast?: boolean;
  style?: string;
  placement?: string;
  children: React.ReactNode;
  className?: string;
}

export interface PopoverProps {
  placement?: string;
  trigger?: string | string[];
  content: React.ReactNode;
  children: React.ReactNode;
  shadow?: boolean;
  modifer?: string;
  rootClose?: boolean;
}

export interface UsageBarProps {
  max?: number;
  now?: number;
  percent?: number;
  unavailableData?: number;
  isByte?: boolean;
  isPercent?: boolean;
  isBulk?: boolean;
  withPercent?: boolean;
  inline?: boolean;
  hideNow?: boolean;
  hideRight?: boolean;
  showZeroMax?: boolean;
  withUnavailable?: boolean;
}

export interface LoaderProps {
  bsSize?: string;
  bsStyle?: string;
  children?: React.ReactNode;
}

export interface TabsProps {
  tabs: any[];
  size?: 'small';
  eventKeyName?: string;
  direction?: 'right';
  onSelect?: SelectCallback;
  activeKey?: string;
  id?: string;
}

export interface StepsProps {
  steps: any[];
  currentStep: number;
  showIcon?: boolean;
  iconSize?: string;
  iconStatus?: string;
}

export interface ModalProps {
  title: string;
  onHide: any;
  onOk?: any;
  show?: boolean;
  style?: React.CSSProperties;
  bsSize?: 'sm' | 'medium' | 'lg' | 'xlarge';
  confirmText?: string;
  okStyle?: string;
  loading?: boolean;
  hideFooter?: boolean;
  hideHeader?: boolean;
}

interface SwitchInput {
  name?: string;
  onChange: any;
}

export interface SwitchProps {
  inline?: boolean;
  input?: SwitchInput;
  name?: string;
  disabled?: boolean;
  handleChange: any;
  checked?: boolean;
  onChange?: any;
}

export interface TimePickerProps {
  placeholder?: string;
  hourStart?: boolean;
  hourEnd?: boolean;
  showSecond?: boolean;
  allowEmpty?: boolean;
  value?: Moment;
  defaultValue?: Moment;
  onChange?: any;
}

export interface TreeData {
  title: string;
  disabled?: boolean;
  key: string;
  children?: TreeData[];
}

export interface TreeProps {
  data: TreeData[];
  checkable?: boolean;
  prefixCls?: string;
  showIcon?: boolean;
  selectable?: boolean;
}

export interface SubMenuProps {
  name?: string;
  title: string;
  children: React.ReactNode;
}

export type DropdownButtonMenuItem =
  | {
      key?: string | number;
      children?: DropdownButtonMenuItem[];
      title: string;
      eventKey?: string;
      'data-action'?: string;
      onClick?: React.MouseEventHandler;
      disabled?: boolean;
    }
  | string;

export interface DefaultDropdownButtonProps {
  componentClass: any;
}
export interface DropdownButtonProps extends DefaultDropdownButtonProps {
  bsStyle?: string;
  id: string;
  onSelect?: SelectCallback;
  onToggle?: (isOpen: boolean) => void;
  bsSize?: Sizes;
  title?: string | React.ReactNode;
  menu?: DropdownButtonMenuItem[];
  children?: React.ReactNode;
  modifer?: string;
  disabled?: boolean;
  dropup?: boolean;
  noCaret?: boolean;
  open?: boolean;
  pullRight?: boolean;
}

export interface NavigationGroup {
  title: string | React.ReactNode;
  icon?: string;
  isFirst?: boolean;
  children?: any[];
  component?: any;
  toggled?: boolean;
}

export interface NavigationProps {
  navGroups: {
    [key: string]: NavigationGroup;
  };
  toggled?: boolean;
  logo?: React.ReactNode | string;
  expandedKeys?: string[];
}

export interface RangePickerProps {
  value?: string[];
  disabled?: boolean;
  lang?: string;
  format?: string;
  onChange?: (data: Moment[], dataString: string[]) => void;
  onOk?: (date: Moment[]) => void;
  onOpenChange?: (open: boolean) => void;
  showDuration?: boolean;
  defaultValue?: Moment[];
}

export interface RangePickerState {
  value: Moment[] | string[];
  open: boolean;
}
export interface AlertProps {
  children: React.ReactNode;
  bsStyle?: string;
  show?: boolean;
  showIcon?: boolean;
  bsClass?: string;
  onDismiss?: () => void;
  dismissDirection?: string;
}

export interface DatePickerDefaultProps {
  disabled?: boolean;
  showTime?: boolean;
}

export interface DatePickerProps extends DatePickerDefaultProps {
  getTime?: (time: string) => void;
  placeholder?: string;
  reset?: boolean;
  defaultValue?: string;
  disabledDate?: (mtime: Moment | undefined) => boolean;
  lang?: 'en' | 'zh';
  disabled?: boolean;
}

export interface PanelProps {
  children?: React.ReactNode;
  bg?: string;
  text?: string;
  className?: string;
  header?: React.ReactNode;
  collapsible?: boolean;
  expanded?: boolean;
}

export interface DropdownDefaultProps {
  id: string;
}

export interface DropdownProps extends DropdownDefaultProps {
  className?: string;
  title?: string;
  children?: React.ReactNode;
  customToggle?: boolean;
  pullRight?: boolean;
}

export interface MenuItemOptions {
  title: string;
  value: string;
}
export interface InputDropdownProps {
  options?: MenuItemOptions[];
  defaultValue?: string;
  value?: string;
  onChange?: SelectCallback;
  input?: any;
  meta?: any;
}
export interface MultiVirtualSelectItem {
  label: string;
  value: number | string;
}
export type VirtualItem =
  | MultiVirtualSelectItem
  | object
  | string
  | number;
export interface SelectCheckItemProps {
  onSelect: Function;
  selected: boolean;
  option: MultiVirtualSelectItem;
  className?: string;
  style?: object;
}
export interface VirtualRowArgs<T> {
  index: number;
  item: T;
  prevItem: T | null;
  nextItem: T | null;
  style: CSS.Properties;
}
export interface VirtualAnchorItem {
  index: number;
  offset: number;
}
export interface VirtualListState {
  startIndex: number;
  endIndex: number;
}
export interface VirtualListDefaultProps<T> {
  height?: number | string;
  data: T[];
  runwayItems?: number;
  runwayItemsOppsite?: number;
  loader?: React.ReactNode;
  placeholder?: React.ReactNode | string;
  noMoreHint?: React.ReactNode | boolean;
  debug?: boolean;
}
export interface VirtualListProps<T> extends VirtualListDefaultProps<T> {
  query?: Query;
  onQueryChange?: (query: Query) => Promise<void>;
  rowHeight?: number | ((item: object) => number);
  rowRenderer: (item: VirtualRowArgs<T>) => React.ReactNode | Element;
  isFetching?: boolean;
  isReloading?: boolean;
  noMore?: boolean;
  totalCount?: number;
  className?: string;
  isEstimate?: boolean;
}

export interface VirtualSelectBoxDefaultProps<T> {
  rowHeight: number;
  isBtn: boolean;
  disabled: boolean;
  placeholder: string;
  query: Query;
  defaultItem: T;
}
export interface VirtualSelectBoxProps<T> extends VirtualSelectBoxDefaultProps<T> {
  fetchData: (
    isReloading: boolean,
    query: Query,
    search?: string,
  ) => Promise<{
    query: Query;
    items: T[];
    totalCount: number;
    error?: string;
  }>;
  item?: T;
  className?: string;
  clear?: boolean;
  onSelect?: Function;
  // onSelect?: (item: T) => void;
  formatOption?: (item: T) => any;
  multi?: boolean;
  value?: any[];
}
export interface VirtualSelectBoxState<T> {
  search: string;
  items: T[];
  query: Query;
  isFetching: boolean;
  totalCount: number;
  isOpen: boolean;
  isReloading: boolean;
  error?: string;
}

export type NotificationItemStatus = 'success' | 'info' | 'process' | 'warning' | 'danger';
export interface NotificationItem {
  id: string;
  status: NotificationItemStatus;
  text: string;
  title?: React.ReactNode | string;
}
export interface NotificationProps extends NotificationItem {
  autoClose?: boolean;
  counter: number;
  intervalMap?: {
    set: (key: string | number, func: () => void, time: number) => void;
    clear: (key: string | number) => void;
    has: (key: string | number) => boolean;
  };
  onDismiss?: Function;
}
export interface NotificationListProps {
  notifications: Map<string, Map<string, NotificationItem>>;
  onDismiss?: Function;
  autoClose?: boolean;
  format?: Function;
}

export interface NotificationListStates {
  expanded?: boolean;
}
