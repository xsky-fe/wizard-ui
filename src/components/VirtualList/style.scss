.VirtualList__wrapper {
  position: relative;
  width: 100%;
  overflow-x: hidden;
  overflow-y: auto;
}
.VirtualList__no-more-hint,
.VirtualList__loader {
  padding-left: 10px
}