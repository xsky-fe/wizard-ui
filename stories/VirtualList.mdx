import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as VirtualListStories from './VirtualList.stories.tsx';

<Meta of={VirtualListStories} />

# VirtualList
当一个资源数据量很大时，一次加载数据会导致浏览器页面卡顿甚至崩溃，VirtualList 支持根据鼠标滚动加载数据，同时销毁超出数据的 `DOM` 结构，保证页面的流畅性。

## 基本用法
- `rowRenderer` 对列表中每行元素进行渲染，可做到动态布局（每列不等高时重新计算并重新渲染）;
- `data` 指要渲染资源数据，可结合 `onQueryChange` 做异步加载
### 空数组
<Canvas of={VirtualListStories.EmptyData} />

### 加载新数据
真实场景中鼠标滚动，页面下滑过程中会加载新的数据。
<Canvas of={VirtualListStories.AsyncEqualHeight} />

### 随机高度
<Canvas of={VirtualListStories.AsyncDynamicHeight} />

## API
<ArgTypes />