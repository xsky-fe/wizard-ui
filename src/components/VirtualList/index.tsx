import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import sum from 'lodash/sum';
import isFunction from 'lodash/isFunction';
import {
  VirtualListState,
  VirtualAnchorItem,
  VirtualListDefaultProps,
  VirtualListProps,
  VirtualItem,
} from '../../interface';
import CSS from 'csstype';
import './style.scss';

const BASIC_STYLES: CSS.Properties = {
  position: 'absolute',
  width: '100%',
};
const RUNWAY_ITEMS = 50;
const RUNWAY_ITEMS_OPPSITE = 20;

function getHeight(el: HTMLDivElement) {
  const styles = window.getComputedStyle(el);
  const height = el.offsetHeight;
  const marginTop = parseFloat(styles.marginTop || '0');
  const marginBottom = parseFloat(styles.marginBottom || '0');
  return height + marginTop + marginBottom;
}

const defaultProps: VirtualListDefaultProps<VirtualItem> = {
  height: '100%',
  data: [],
  runwayItems: RUNWAY_ITEMS,
  runwayItemsOppsite: RUNWAY_ITEMS_OPPSITE,
  loader: <p className="VirtualList__loader">数据加载中</p>,
  placeholder: <p className="VirtualList__placeholder">暂无数据</p>,
  noMoreHint: <p className="VirtualList__no-more-hint">没有更多信息</p>,
  debug: true,
};

export default class VirtualList<T> extends React.Component<VirtualListProps<T>, VirtualListState> {
  static propTypes = {
    /** 行高 */
    rowHeight: PropTypes.oneOfType([PropTypes.func, PropTypes.number]).isRequired,
    /** 渲染 row UI */
    rowRenderer: PropTypes.func.isRequired,
    /** 可见高度 */
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** 数据 */
    data: PropTypes.array,
    /** 是否开启计算 */
    isEstimate: PropTypes.bool,
    /** 监听 query 改变 */
    onQueryChange: PropTypes.func,
    /** 确定异步数据的 offset 和 limit */
    query: PropTypes.object,
    /** 总数 */
    totalCount: PropTypes.number,
    /** 是否在获取数据 */
    isFetching: PropTypes.bool,
    /** 行高 */
    runwayItems: PropTypes.number,
    /** 行高 */
    runwayItemsOppsite: PropTypes.number,
    /** 数据加载中时展示 */
    loader: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    /** 默认展示文本 */
    placeholder: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    /** 是否展示没有多余的数据 */
    noMore: PropTypes.bool,
    /** 没有需要加载的数据时展示 */
    noMoreHint: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    /** Debug */
    debug: PropTypes.bool,
  };
  static defaultProps = defaultProps;
  holder: React.RefObject<HTMLDivElement>;
  list: React.RefObject<HTMLDivElement>;
  heightCache: number[];
  totalHeight: number;
  anchorItem: VirtualAnchorItem;
  anchorScrollTop: number;
  constructor(props: VirtualListProps<T>) {
    super(props);
    this.holder = React.createRef();
    this.list = React.createRef();
    this.heightCache = [];
    this.totalHeight = 0;
    this.anchorItem = {
      index: 0,
      offset: 0,
    };
    this.anchorScrollTop = 0;
    this.state = {
      startIndex: 0,
      endIndex: 0,
    };
  }

  get wrapperHeight() {
    const { height } = this.props;
    if (typeof height === 'number') {
      return height;
    }
    if (this.holder.current && this.holder.current.offsetHeight) {
      return this.holder.current.offsetHeight;
    }
    return 450;
  }
  get noMore() {
    const { noMore, totalCount, data } = this.props;
    if (noMore) {
      return true;
    }
    if (totalCount === data.length) {
      return true;
    }
    return false;
  }

  componentDidMount() {
    const { data } = this.props;
    this.handleResize(data);
    window.addEventListener('resize', this.resizeHandler);
  }
  componentDidUpdate(prevProps: VirtualListProps<T>) {
    const { isEstimate, debug, data, isReloading } = this.props;
    const { startIndex } = this.state;

    if (prevProps.data !== data) {
      this.handleResize(data, false);
    }
    if (!prevProps.isReloading && isReloading) {
      this.reset();
      if (this.list.current) {
        this.list.current.scrollTop = 0;
      }
      this.setState({
        startIndex: 0,
        endIndex: 0,
      });
    }

    let needRender = false;
    if (isEstimate && this.list.current && this.list.current.querySelectorAll) {
      const rows: NodeListOf<HTMLDivElement> = this.list.current.querySelectorAll(
        '.VirtualList > *',
      );
      for (let i = 0; i < rows.length; i++) {
        const row = rows[i];
        const realHeight = getHeight(row);
        const index = startIndex + i;
        if (realHeight !== this.heightCache[index]) {
          if (debug && process && process.env.NODE_ENV === 'development') {
            console.warn(
              `Index ${index} estimate height is ${this.heightCache[index]}, real height is ${realHeight}.`,
            );
          }
          this.heightCache[index] = realHeight;
          needRender = true;
        }
      }
      if (needRender) {
        this.totalHeight = sum(this.heightCache);
        this.forceUpdate();
      }
    }
  }
  componentWillUnmount() {
    window.removeEventListener('resize', this.resizeHandler);
  }

  reset() {
    this.heightCache = [];
    this.anchorItem = {
      index: 0,
      offset: 0,
    };
    this.anchorScrollTop = 0;
  }

  // resizeHandler is a wrapper of this.handleResize
  // only used for addEventListener and removeEventListener
  resizeHandler = () => {
    this.handleResize(this.props.data);
  };
  handleResize = (data: T[], flushCache?: boolean) => {
    this.recomputeRowHeight(data, flushCache);
    this.handleScroll();
  };
  correctAnchor = () => {
    const { index, offset } = this.anchorItem;
    this.anchorScrollTop = sum(this.heightCache.slice(0, index)) + offset;
  };
  recomputeRowHeight = (nextData: T[], flushCache: boolean = true) => {
    if (flushCache) {
      this.heightCache = [];
    }
    const { rowHeight } = this.props;
    const data = nextData || this.props.data;
    const count = data.length;

    for (let i = this.heightCache.length; i < count; i++) {
      const item = data[i];
      const prevItem = i > 0 ? data[i - 1] : null;
      const nextItem = i < data.length - 1 ? data[i + 1] : null;
      const height = isFunction(rowHeight)
        ? rowHeight({
            index: i,
            item,
            prevItem,
            nextItem,
          })
        : rowHeight;
      if (height) {
        this.heightCache.push(height);
      }
    }
    this.totalHeight = sum(this.heightCache);
    this.forceUpdate();
  };

  handleScroll = () => {
    const { runwayItems = RUNWAY_ITEMS, runwayItemsOppsite = RUNWAY_ITEMS_OPPSITE } = this.props;
    const { scrollTop = 0, offsetHeight = 0 } = this.list.current || {};
    const delta = scrollTop - this.anchorScrollTop;

    this.anchorItem = scrollTop === 0 ? { index: 0, offset: 0 } : this.calculateAnchoredItem(delta);
    this.anchorScrollTop = scrollTop;

    const lastVisibleItem = this.calculateAnchoredItem(offsetHeight);
    let startIndex;
    let endIndex;
    if (delta < 0) {
      startIndex = Math.max(0, this.anchorItem.index - runwayItems);
      endIndex = Math.min(lastVisibleItem.index + runwayItemsOppsite, this.heightCache.length);
    } else {
      startIndex = Math.max(0, this.anchorItem.index - runwayItemsOppsite);
      endIndex = Math.min(lastVisibleItem.index + runwayItems, this.heightCache.length);
    }
    this.setState(
      {
        startIndex,
        endIndex,
      },
      this.tryToFetchData,
    );
  };

  calculateAnchoredItem = (delta: number) => {
    const { data } = this.props;

    if (delta === 0) return this.anchorItem;
    let { index, offset: initialOffset } = this.anchorItem;
    delta += initialOffset;

    if (delta < 0) {
      while (delta < 0 && index > 0 && this.heightCache[index] !== undefined) {
        delta += this.heightCache[index - 1];
        index--;
      }
    } else {
      while (
        delta > 0 &&
        index < data.length &&
        this.heightCache[index] !== undefined &&
        this.heightCache[index] < delta
      ) {
        delta -= this.heightCache[index];
        index++;
      }
    }

    return {
      index,
      offset: delta,
    };
  };

  attachItems = () => {
    this.correctAnchor();
    const { data } = this.props;
    const { startIndex, endIndex } = this.state;
    if (this.heightCache.length === 0) return [];

    let items = [];
    let { index, offset } = this.anchorItem;
    let currentPosition = this.anchorScrollTop - offset;
    while (index > startIndex) {
      currentPosition -= this.heightCache[index - 1];
      index--;
    }
    while (index < startIndex) {
      currentPosition += this.heightCache[index];
      index++;
    }

    for (let i = startIndex; i < endIndex; i++) {
      const item = data[i];
      const prevItem = i > 0 ? data[i - 1] : null;
      const nextItem = i < data.length - 1 ? data[i + 1] : null;
      items.push({
        index: i,
        item,
        prevItem,
        nextItem,
        style: {
          ...BASIC_STYLES,
          transform: `translateY(${currentPosition}px)`,
        },
      });
      currentPosition += this.heightCache[i];
    }
    return items;
  };

  tryToFetchData = () => {
    const { data, query, onQueryChange, isFetching } = this.props;
    const { endIndex } = this.state;
    if (isFetching) return;
    if (endIndex < data.length) return;
    if (this.noMore) return;
    if (onQueryChange && query) {
      const { offset = 0, limit = 0 } = query;
      onQueryChange({
        ...query,
        offset: offset + limit,
        limit,
      });
    }
  };

  render() {
    const {
      rowRenderer,
      data,
      isFetching,
      isReloading,
      height,
      loader,
      placeholder,
      noMoreHint,
      className,
    } = this.props;
    const noData = data.length === 0;
    return (
      <div className="VirtualList__holder" ref={this.holder} style={{ height }}>
        <div
          className="VirtualList__wrapper"
          style={{ height: this.wrapperHeight }}
          onScroll={this.handleScroll}
          ref={this.list}
        >
          {!isFetching && noData && placeholder}
          {!isReloading && !noData && (
            <div
              className={classnames('VirtualList', className)}
              style={{
                width: '100%',
                height: this.totalHeight,
              }}
            >
              {this.attachItems().map(rowRenderer)}
            </div>
          )}
          {isFetching && loader}
          {!isFetching && !noData && this.noMore && noMoreHint}
        </div>
      </div>
    );
  }
}
