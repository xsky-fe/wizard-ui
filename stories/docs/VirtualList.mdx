import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as VirtualListStories from '../VirtualList.stories.tsx';

<Meta of={VirtualListStories} />

# VirtualList
<em><strong>(wizard-ui自定义组件)</strong></em>
<p>2023-6-14 [chensheng](https://github.xsky.com/chen-sheng)</p>
  当一个资源数据量很大时，一次加载数据会导致浏览器页面卡顿甚至崩溃，VirtualList 支持根据鼠标滚动加载数据，同时销毁超出数据的 `DOM` 结构，保证页面的流畅性。

## 基本用法
- `rowRenderer` 对列表中每行元素进行渲染，可做到动态布局（每列不等高时重新计算并重新渲染）;
- `data` 指要渲染资源数据，可结合 `onQueryChange` 做异步加载
### 空数组
<Canvas of={VirtualListStories.EmptyData} />

### 加载新数据
真实场景中鼠标滚动，页面下滑过程中会加载新的数据。
<Canvas of={VirtualListStories.AsyncEqualHeight} source={{code:`import React from 'react';
import { VirtualList } from '../../src';
import { getMockDatas } from '../../src/utils';
import { get } from 'lodash';
import { Query, VirtualRowArgs, VirtualItem } from '../../src/interface';

const resName = 'list';
function getDatas(query: Query) {
  return getMockDatas(query, 180, resName);
}

export const rowRenderer = (i: VirtualRowArgs<VirtualItem>) => (
  <div className="list-item" style={i.style} key={i.index}>
    {resName}-{i.index + 1}
  </div>
);

const rowRendererRandomHeight = (i: VirtualRowArgs<VirtualItem>) => (
  <div className="list-item" style={i.style} key={i.index}>
    <div style={{ height: \`\${35 + (i.index % 5) * 4}px\` }}>
      {resName}-{i.index + 1}
    </div>
  </div>
);

export default (props: { random?: boolean }) => {
  const [fetching, setFetch] = React.useState(false);
  const [datas, setDatas] = React.useState([]);
  const [totalCount, setTotalCount] = React.useState(0);
  const [count, setCount] = React.useState(0);
  React.useEffect(() => {
    handleQueryChange({
      limit: 30,
      offset: 0,
    });
  }, []);
  React.useEffect(() => {
    const existLists = document.querySelectorAll('.VirtualList > *');
    setCount(existLists ? existLists.length : 0);
  }, [datas]);
  const handleQueryChange = async (query: Query) => {
    setFetch(true);
    const actionResult: any = await getDatas(query);
    setFetch(false);
    const totalCount = get(actionResult, 'response.paging.totalCount');
    const lists = actionResult.response.lists;
    setTotalCount(totalCount);
    setDatas(datas.concat(lists));
  };
  return (
    <div>
      <p>
        总共 {totalCount}, 获取了 {datas.length} 条数据。
      </p>
      <p>
        VirtualList 会在增加 list 条数的同时，销毁超出的 UI，目前渲染出来的 list 条数是： {count}
      </p>
      <div className="box">
        <VirtualList
          rowHeight={35}
          rowRenderer={props.random ? rowRendererRandomHeight : rowRenderer}
          height={210}
          data={datas}
          query={{
            limit: 30,
            offset: 0,
          }}
          onQueryChange={handleQueryChange}
          totalCount={totalCount}
          isFetching={fetching}
          noMore={datas.length === totalCount}
          isEstimate={props.random}
        />
      </div>
    </div>
  );
};`}}/>

### 随机高度
<Canvas of={VirtualListStories.AsyncDynamicHeight} source={{code:'见上'}}/>

## API
<ArgTypes />