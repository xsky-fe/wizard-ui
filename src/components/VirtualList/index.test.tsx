import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import VirtualList from './index';
import { VirtualRowArgs, VirtualItem } from '../../interface';
import AsyncVirtualList from '../../../stories/demos/AsyncVirtualList';

function createNodeMock(element: React.ReactElement) {
  if (element.type === 'div') {
    return {
      scrollTop: 0,
      offsetHeight: 100,
    };
  }
  return null;
}
const snapOptions = { createNodeMock };

const rowRenderer = (i: VirtualRowArgs<VirtualItem>) => <div key={i.index}>item</div>;

describe('VirtualList', () => {
  it('render without crush', () => {
    const list = shallow(<VirtualList rowHeight={10} rowRenderer={rowRenderer} />);
    expect(list.hasClass('VirtualList__holder')).toBeTruthy();
  });
  it('show no data placeholder when data array is empty', () => {
    const list = renderer
      .create(
        <VirtualList
          rowHeight={10}
          rowRenderer={rowRenderer}
          data={[]}
          isFetching={false}
          isReloading={false}
          totalCount={0}
        />,
        snapOptions,
      )
      .toJSON();
    expect(list).toMatchSnapshot();
  });

  it('show list and loader when fetching more', () => {
    const list = renderer
      .create(
        <VirtualList
          rowHeight={10}
          rowRenderer={rowRenderer}
          data={[1]}
          isFetching={true}
          isReloading={false}
          totalCount={1}
        />,
        snapOptions,
      )
      .toJSON();
    expect(list).toMatchSnapshot();
  });

  it('only show loader when reloading', () => {
    const list = renderer
      .create(
        <VirtualList
          rowHeight={10}
          rowRenderer={rowRenderer}
          data={[]}
          isFetching={true}
          isReloading={true}
          totalCount={0}
        />,
        snapOptions,
      )
      .toJSON();
    expect(list).toMatchSnapshot();
  });

  it('show no more data hint when all data fetched', () => {
    const list = renderer
      .create(
        <VirtualList
          rowHeight={10}
          rowRenderer={rowRenderer}
          data={[1]}
          isFetching={false}
          isReloading={false}
          totalCount={1}
        />,
        snapOptions,
      )
      .toJSON();
    expect(list).toMatchSnapshot();
  });

  it('scrolling equal row height', () => {
    const list = renderer.create(<AsyncVirtualList />, snapOptions).toJSON();
    expect(list).toMatchSnapshot();
  });
  it('scrolling dynamic row height', () => {
    const list = renderer.create(<AsyncVirtualList random />, snapOptions).toJSON();
    expect(list).toMatchSnapshot();
  });
});
