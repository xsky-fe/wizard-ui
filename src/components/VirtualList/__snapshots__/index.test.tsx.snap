// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualList only show loader when reloading 1`] = `
<div
  className="VirtualList__holder"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="VirtualList__wrapper"
    onScroll={[Function]}
    style={
      Object {
        "height": 100,
      }
    }
  >
    <p
      className="VirtualList__loader"
    >
      数据加载中
    </p>
  </div>
</div>
`;

exports[`VirtualList scrolling dynamic row height 1`] = `
<div>
  <p>
    总共 
    0
    , 获取了 
    0
     条数据。
  </p>
  <p>
    VirtualList 会在增加 list 条数的同时，销毁超出的 UI，目前渲染出来的 list 条数是： 
    0
  </p>
  <div
    className="box"
  >
    <div
      className="VirtualList__holder"
      style={
        Object {
          "height": 210,
        }
      }
    >
      <div
        className="VirtualList__wrapper"
        onScroll={[Function]}
        style={
          Object {
            "height": 210,
          }
        }
      >
        <p
          className="VirtualList__loader"
        >
          数据加载中
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtualList scrolling equal row height 1`] = `
<div>
  <p>
    总共 
    0
    , 获取了 
    0
     条数据。
  </p>
  <p>
    VirtualList 会在增加 list 条数的同时，销毁超出的 UI，目前渲染出来的 list 条数是： 
    0
  </p>
  <div
    className="box"
  >
    <div
      className="VirtualList__holder"
      style={
        Object {
          "height": 210,
        }
      }
    >
      <div
        className="VirtualList__wrapper"
        onScroll={[Function]}
        style={
          Object {
            "height": 210,
          }
        }
      >
        <p
          className="VirtualList__loader"
        >
          数据加载中
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtualList show list and loader when fetching more 1`] = `
<div
  className="VirtualList__holder"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="VirtualList__wrapper"
    onScroll={[Function]}
    style={
      Object {
        "height": 100,
      }
    }
  >
    <div
      className="VirtualList"
      style={
        Object {
          "height": 10,
          "width": "100%",
        }
      }
    >
      <div>
        item
      </div>
    </div>
    <p
      className="VirtualList__loader"
    >
      数据加载中
    </p>
  </div>
</div>
`;

exports[`VirtualList show no data placeholder when data array is empty 1`] = `
<div
  className="VirtualList__holder"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="VirtualList__wrapper"
    onScroll={[Function]}
    style={
      Object {
        "height": 100,
      }
    }
  >
    <p
      className="VirtualList__placeholder"
    >
      暂无数据
    </p>
  </div>
</div>
`;

exports[`VirtualList show no more data hint when all data fetched 1`] = `
<div
  className="VirtualList__holder"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="VirtualList__wrapper"
    onScroll={[Function]}
    style={
      Object {
        "height": 100,
      }
    }
  >
    <div
      className="VirtualList"
      style={
        Object {
          "height": 10,
          "width": "100%",
        }
      }
    >
      <div>
        item
      </div>
    </div>
    <p
      className="VirtualList__no-more-hint"
    >
      没有更多信息
    </p>
  </div>
</div>
`;
